middlewares/cors.go

```diff
@@ -0,0 +1,179 @@
+package middlewares
+
+import (
+	"github.com/savsgio/atreugo/v10"
+	"github.com/savsgio/go-logger"
+
+	"strconv"
+	"strings"
+)
+
+type CorsOptions struct {
+	AllowedOrigins   []string
+	AllowedHeaders   []string
+	AllowMaxAge      int
+	AllowedMethods   []string
+	ExposedHeaders   []string
+	AllowCredentials bool
+}
+
+type CorsHandler struct {
+	maxAge            int
+	allowCredentials  bool
+	allowedOriginsAll bool
+	allowedHeadersAll bool
+	allowedOrigins    []string
+	allowedHeaders    []string
+	allowedMethods    []string
+	exposedHeaders    []string
+}
+
+var defaultOptions = &CorsOptions{
+	AllowedOrigins: []string{"*"},
+	AllowedMethods: []string{"GET", "POST"},
+	AllowedHeaders: []string{"Origin", "Accept", "Content-Type"},
+}
+
+func DefaultCors() *CorsHandler {
+	return NewCorsMiddleware(*defaultOptions)
+}
+
+func NewCorsMiddleware(options CorsOptions) *CorsHandler {
+	cors := &CorsHandler{
+		allowedOrigins:   options.AllowedOrigins,
+		allowedHeaders:   options.AllowedHeaders,
+		allowCredentials: options.AllowCredentials,
+		allowedMethods:   options.AllowedMethods,
+		exposedHeaders:   options.ExposedHeaders,
+		maxAge:           options.AllowMaxAge,
+	}
+
+	if len(cors.allowedOrigins) == 0 {
+		cors.allowedOrigins = defaultOptions.AllowedOrigins
+		cors.allowedOriginsAll = true
+	} else {
+		for _, v := range options.AllowedOrigins {
+			if v == "*" {
+				cors.allowedOrigins = defaultOptions.AllowedOrigins
+				cors.allowedOriginsAll = true
+				break
+			}
+		}
+	}
+
+	if len(cors.allowedHeaders) == 0 {
+		cors.allowedHeaders = defaultOptions.AllowedHeaders
+		cors.allowedHeadersAll = true
+	} else {
+		for _, v := range options.AllowedHeaders {
+			if v == "*" {
+				cors.allowedHeadersAll = true
+				break
+			}
+		}
+	}
+
+	if len(cors.allowedMethods) == 0 {
+		cors.allowedMethods = defaultOptions.AllowedMethods
+	}
+
+	return cors
+}
+
+func (c *CorsHandler) CorsMiddleware(ctx *atreugo.RequestCtx) {
+	c.handlePreflight(ctx)
+}
+
+func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) {
```

Comment:
The checks are not necesary because server just sets the headers but shouldn't makes checks, it's the browser or client that make it.

I suggest you to refactor with this code:
```go
func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) error {
	originHeader := string(ctx.Request.Header.Peek("Origin"))

	if !c.isAllowedOrigin(originHeader) {
		return ctx.Next()
	}

	ctx.Response.Header.Set("Access-Control-Allow-Origin", originHeader)

	if c.allowCredentials {
		ctx.Response.Header.Set("Access-Control-Allow-Credentials", "true")
	}

	varyHeader := ctx.Response.Header.Peek("Vary")
	if len(varyHeader) > 0 {
		varyHeader = append(varyHeader, ", "...)
	}

	varyHeader = append(varyHeader, "Origin"...)

	ctx.Response.Header.SetBytesV("Vary", varyHeader)

	method := string(ctx.Request.Header.Method())
	if method != fasthttp.MethodOptions {
		return ctx.Next()
	}

	if len(c.allowedMethods) > 0 {
		ctx.Response.Header.Set("Access-Control-Allow-Methods", strings.Join(c.allowedMethods, ", "))
	}

	if len(c.allowedHeaders) > 0 {
		ctx.Response.Header.Set("Access-Control-Allow-Headers", strings.Join(c.allowedHeaders, ", "))
	}

	if c.maxAge > 0 {
		ctx.Response.Header.Set("Access-Control-Max-Age", strconv.Itoa(c.maxAge))
	}

	return ctx.Next()
}
```